catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceQueryService.java

```diff
@@ -0,0 +1,386 @@
+            scheduler.deleteJob(jobDetail.getKey());
+            scheduler.scheduleJob(jobDetail, new CronString(cronString, TRIGGER_NAME).get());
+        } catch (SchedulerException e) {
+            LOGGER.warn("unable to set the cron string: cron=[{}]", cronString, e);
+        }
+    }
+
+    /**
+     * @param queryTimeoutMinutes minutes (must be non-null)
+     */
+    @SuppressWarnings("unused")
+    public void setQueryTimeoutMinutes(Long queryTimeoutMinutes) {
+        notNull(queryTimeoutMinutes, "queryTimeoutMinutes must be non-null");
+        this.queryTimeoutMinutes = queryTimeoutMinutes;
+    }
+
+    /**
+     * Main entry point, should be called by a scheduler.
+     */
+    public void run() {
+
+        LOGGER.debug("running workspace query service");
+
+        Map<String, Pair<WorkspaceMetacardImpl, List<QueryMetacardImpl>>> queryMetacards =
+                workspaceService.getQueryMetacards();
+
+        LOGGER.debug("queryMetacards: size={}", queryMetacards.size());
+
+        List<WorkspaceTask> workspaceTasks = createWorkspaceTasks(queryMetacards);
+
+        LOGGER.debug("workspaceTasks: size={}", workspaceTasks.size());
+
+        Map<String, Pair<WorkspaceMetacardImpl, Long>> results = executeWorkspaceTasks(
+                workspaceTasks,
+                queryTimeoutMinutes,
+                TimeUnit.MINUTES);
+
+        LOGGER.debug("results: {}", results);
+
+        queryUpdateSubscriber.notify(results);
+
+    }
+
+    private Map<String, Pair<WorkspaceMetacardImpl, Long>> executeWorkspaceTasks(
+            List<WorkspaceTask> workspaceTasks, long timeout, TimeUnit timeoutUnit) {
+        Map<String, Pair<WorkspaceMetacardImpl, Long>> results = new ConcurrentHashMap<>();
+
+        workspaceTasks.stream()
+                .map(ForkJoinPool.commonPool()::submit)
+                .map(task -> getTaskResult(task, timeout, timeoutUnit))
+                .filter(Objects::nonNull)
+                .forEach(pair -> results.put(pair.getLeft()
+                        .getId(), new ImmutablePair<>(pair.getLeft(), pair.getRight())));
+
+        return results;
+    }
+
+    private Pair<WorkspaceMetacardImpl, Long> getTaskResult(
+            ForkJoinTask<Pair<WorkspaceMetacardImpl, Long>> workspaceTask, long timeout,
+            TimeUnit timeoutUnit) {
+        try {
+            return workspaceTask.get(timeout, timeoutUnit);
+        } catch (TimeoutException e) {
+            LOGGER.warn("Timeout", e);
+        } catch (ExecutionException | InterruptedException e) {
+            LOGGER.warn("ForkJoinPool error", e);
+        }
+        return null;
+    }
+
+    private List<WorkspaceTask> createWorkspaceTasks(
+            Map<String, Pair<WorkspaceMetacardImpl, List<QueryMetacardImpl>>> queryMetacards) {
+
+        List<WorkspaceTask> workspaceTasks = new ArrayList<>();
+
+        for (Map.Entry<String, Pair<WorkspaceMetacardImpl, List<QueryMetacardImpl>>> workspaceQueryEntry : queryMetacards.entrySet()) {
```

Comment:
Since you're only using the entry values, you could just iterate over `queryMetacards.values()`.